algorithms/kernel/dbscan/oneapi/cl_kernels/dbscan_cl_kernels.cl

```diff
@@ -0,0 +1,182 @@
+        }
+    }
+
+    __kernel void queue_block_distances(__global const algorithmFPType * data, __global const int * queue, __global algorithmFPType * dist,
+                                        int queue_begin, int queue_size, int power, int dim, int num_points) {
+        const int queue_pos = get_global_id(0);
+        if (queue_pos >= queue_size) return;
+        const int id                        = queue[queue_begin + queue_pos];
+        const int group_num                 = get_num_sub_groups();
+        const int group_id                  = get_sub_group_id();
+        const int group_size                = get_sub_group_size();
+        const int local_id                  = get_sub_group_local_id();
+        __global algorithmFPType * cur_dist = &dist[queue_pos * num_points];
+        for (int j = group_id; j < num_points; j += group_num)
+        {
+            algorithmFPType sum = 0.0;
+            for (int i = local_id; i < dim; i += group_size)
+            {
+                algorithmFPType val = fabs(data[id * dim + i] - data[j * dim + i]);
+                algorithmFPType dm  = 1.0;
+                for (int j = 0; j < power; j++)
+                {
+                    dm *= val;
+                }
+                sum += dm;
+            }
+            algorithmFPType ret = sub_group_reduce_add(sum);
+            if (local_id == 0)
+            {
+                cur_dist[j] = ret;
+            }
+        }
+    }
+
+    __kernel void count_neighbors(__global const int * assignments, __global const algorithmFPType * data, int row_id, int chunk_offset,
+                                  int chunk_size, int num_points, algorithmFPType eps, __global const int * queue, __global int * counters,
+                                  __global int * undefCounters) {
+        const int index           = get_global_id(0) * get_num_sub_groups() + get_sub_group_id();
+        const int offset          = index * chunk_size;
+        const int number          = num_points - offset < chunk_size ? num_points - offset : chunk_size;
+        row_id                    = chunk_offset < 0 ? row_id : queue[row_id];
+        const int distance_offset = offset + (chunk_offset < 0 ? 0 : chunk_offset);
+        if (index >= num_points) return;
+        const int size                         = get_sub_group_size();
+        const int id                           = get_sub_group_local_id();
+        __global const algorithmFPType * input = &data[distance_offset];
+        __global const int * assigned          = &assignments[offset];
+        int count                              = 0;
+        int new_count                          = 0;
+        for (int i = id; i < number; i += size)
+        {
+            int nbr_flag     = input[i] <= eps ? 1 : 0;
+            int new_nbr_flag = nbr_flag > 0 && assigned[i] == _UNDEFINED_ && (i + offset != row_id) ? 1 : 0;
+            count += nbr_flag;
+            new_count += new_nbr_flag;
```

Comment:
Totally not clear. What is the "new" neighbor? Why "undefined" label is distinguished from others? Why we differ `count` and `new_count`?